File: 4th Semester/Design and Programming Environments/labs/3rd lab/Java/src/main/java/ro/mpp2024/service/Service.java
package ro.mpp2024.service;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ro.mpp2024.domain.Game;
import ro.mpp2024.domain.Purchase;
import ro.mpp2024.domain.User;
import ro.mpp2024.repository.*;

import ro.mpp2024.repository.RepositoryUser;

import java.util.ArrayList;
import java.util.Optional;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;



@Component
public class Service {
    private final RepositoryGame game_repo;
    private final RepositoryUser user_repo;
    private final RepositoryPurchase purchase_repo;

    @Autowired
    public Service(RepositoryGame game_repo, RepositoryUser user_repo, RepositoryPurchase purchase_repo) {
        this.game_repo = game_repo;
        this.user_repo = user_repo;
        this.purchase_repo = purchase_repo;
    }

    //loginul
    public Optional<User> login(String username, String password){
        for (User user : user_repo.findAll()){
            if(user.getUsername().equals(username) && user.getPassword().equals(password)){
                return Optional.of(user);
            }
        }
        return Optional.empty();
        //return user_repo.findAll().stream()
        //                .filter(user -> user.getUsername().equals(username) && user.getPassword().equals(password))
        //                .findFirst();
    }

    public List<Purchase> find_by_client(String client){
        List<Purchase> purchases = new ArrayList<>();
        purchases=purchase_repo.findByClientOrderedBySeats(client);
        return purchases;
    }

    public void signup(String name, String username, String password){
        for(User user : user_repo.findAll()){
            if(user.getUsername().equals(username)){
                throw new IllegalArgumentException("Username already exists.");            }
        }
        validatePassword(password);

        //Optional<User> existingUser = user_repo.findAll().stream()
        //                .filter(user -> user.getUsername().equals(username))
        //                .findFirst();
        User newUser = new User(name, username, password);
        user_repo.save(newUser);
    }
    private void validatePassword(String password) {
        if (password == null) {
            throw new IllegalArgumentException("Password is required");
        }
        if (password.length() < 8) {
            throw new IllegalArgumentException("Password must be at least 8 characters");
        }
        if (!Pattern.compile("[!@#$%^&]").matcher(password).find()) {
            throw new IllegalArgumentException("Password must contain at least one special character (!@#$%^&)");
        }
        if (!Pattern.compile("[a-z]").matcher(password).find()) {
            throw new IllegalArgumentException("Password must contain at least one lowercase letter");
        }
        if (!Pattern.compile("[A-Z]").matcher(password).find()) {
            throw new IllegalArgumentException("Password must contain at least one uppercase letter");
        }
        if (!Pattern.compile("[0-9]").matcher(password).find()) {
            throw new IllegalArgumentException("Password must contain at least one digit");
        }
    }
    public List<Game> showcase_games_by_type(String type){
        List<Game> games = new ArrayList<>();
        games=game_repo.findByTypeOrderedByDate(type);
        return games;
    }
    public void add_purchase(Purchase purchase){
        Game game = game_repo.findOne(purchase.getGame()).orElseThrow(() -> new IllegalArgumentException("game not found."));
        update_seats(game, purchase.getSeats());
        purchase_repo.save(purchase);
    }
    public void update_seats(Game game, int seats){
        if(seats<=0)
            throw new IllegalArgumentException("invalid number of seats!");
        if(game.getSeats()<seats)
            throw new IllegalArgumentException("not enough number of seats!");
        game.setSeats(game.getSeats()-seats);
        game_repo.update(game);
    }
}

File: 4th Semester/Design and Programming Environments/labs/3rd lab/C#/basket2/service/Service.cs
using System.Reflection;
using basket2.domain;
using basket2.repository;
using log4net;

namespace basket2.service;
using System.Text.RegularExpressions;

public class Service
{
    private static readonly ILog log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);

    private GameRepository game_repo;
    private PurchaseRepository purchase_repo;
    private UserRepository user_repo;

    public Service(GameRepository game_repo, PurchaseRepository purchase_repo, UserRepository user_repo)
    {
        this.game_repo = game_repo;
        this.purchase_repo = purchase_repo;
        this.user_repo = user_repo;
    }

    public User? Login(string username, string password)
    {
        
        return user_repo.findAll().FirstOrDefault(user=>user.username == username && user.password == password);
    }

    public List<Purchase> findByClientId(string client)
    {
        return purchase_repo.findByClientOrderedBySeats(client);
    }

    public void Signup(string name, string username, string password)
    {
        if (user_repo.findAll().Any(user => user.username == username))
        {
            throw new ArgumentException("Username already exists");
        }

        ValidatePassword(password);
        var new_user=new User(name, username, password);
        user_repo.Save(new_user);
    }

    private void ValidatePassword(string password)
    {
        if(string.IsNullOrEmpty(password))
            throw new ArgumentException("Password is required!");
        if(password.Length < 8)
            throw new ArgumentException("Password must be at least 8 characters long!");
        if(!Regex.IsMatch(password, "[!@#$%^&]"))
            throw new ArgumentException("Password must contain at least one special character (!@#$%^&)!");
        if(!Regex.IsMatch(password, "[a-z]"))
            throw new ArgumentException("Password must contain at least one lowercase letter");
        if(!Regex.IsMatch(password, "[A-Z]"))
            throw new ArgumentException("Password must contain at least one uppercase letter");
        if(!Regex.IsMatch(password, "[0-9]"))
            throw new ArgumentException("Password must contain at least one digit");
    }

    public List<Game> ShowcaseGamesByType(string gameType)
    {
        var games = game_repo.findByTypeOrderedByDate(gameType);
        return new List<Game>(games);  // Returnează o copie nouă
    }




    public void AddPurchase(Purchase purchase)
    {
        var game=game_repo.findOne(purchase.game) ?? throw new ArgumentException("Game not found");
        UpdateSeats(game, purchase.seats);
        purchase_repo.Save(purchase);
    }

    public void UpdateSeats(Game game, int seats)
    {
        if(seats<=0)
            throw new ArgumentException("Invalid number of seats!");
        if(game.Seats!<seats)
            throw new ArgumentException("Not enough number of seats!");
        game.Seats -= seats;
        game_repo.Update(game);
    }
}
